File: sdk/python/smoke_test.py
"""Self-contained SDK smoke test: builds atlas-server, runs it on a temp store,
exercises the full lifecycle through the SDK, and tears down.

    python3 sdk/python/smoke_test.py
"""
import os
import pathlib
import shutil
import socket
import subprocess
import tempfile
import time

from atlas import AtlasClient, AtlasError, Decision


def free_port() -> int:
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


def main() -> None:
    root = pathlib.Path(__file__).resolve().parents[2]
    workdir = tempfile.mkdtemp(prefix="atlas-sdk-")
    binp = os.path.join(workdir, "atlas-server")
    subprocess.run(["go", "build", "-o", binp, "./cmd/atlas-server"], cwd=root, check=True)

    port = free_port()
    proc = subprocess.Popen(
        [binp, "-addr", f"127.0.0.1:{port}",
         "-store", f"{workdir}/state.json", "-key", f"{workdir}/authority.key"],
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
    )
    try:
        atlas = AtlasClient(f"http://127.0.0.1:{port}")
        for _ in range(60):
            if atlas.health():
                break
            time.sleep(0.1)
        else:
            raise SystemExit("server did not become healthy")

        p, d = "spiffe://domain-a.test/workload/api", "spiffe://domain-b.test/agent/worker"
        grant = atlas.issue(p, d, ["read:orders", "write:audit"])
        assert atlas.verify(grant.record).decision == Decision.ACCEPT, "fresh grant should accept"

        try:
            atlas.issue(p, d, ["delete:everything"])
            raise AssertionError("over-scope should have been refused")
        except AtlasError:
            pass  # expected

        atlas.revoke(grant.instance)
        r = atlas.verify(grant.record)
        assert r.decision == Decision.REJECT, f"revoked grant should reject, got {r.decision}"
        assert "RevokedObservable" in r.causes, r.causes

        assert atlas.stats()["issued"] >= 1
        assert len(atlas.delegations()) >= 1
        assert atlas.graph()["edges"], "graph should have edges"

        print("PASS: atlas python sdk smoke (issue, attenuation, offline verify, revoke)")
    finally:
        proc.terminate()
        proc.wait()
        shutil.rmtree(workdir, ignore_errors=True)


if __name__ == "__main__":
    main()

File: sdk/typescript/smoke_test.ts
/**
 * Live smoke test for the Atlas TypeScript SDK.
 *
 * Requires a running server:  go run ./cmd/atlas-server
 * Run (Node 22.6+, no deps):  node --experimental-strip-types sdk/typescript/smoke_test.ts
 *   or with a base URL:        ATLAS_URL=http://127.0.0.1:8087 node --experimental-strip-types ...
 *
 * Exercises the full alpha-path: issue -> verify(accept) -> revoke ->
 * verify(reject/RevokedObservable), plus /health and /readyz.
 */
import { AtlasClient, Decision } from "./atlas.ts";

const url = process.env.ATLAS_URL ?? "http://127.0.0.1:8087";
const atlas = new AtlasClient(url, { apiKey: process.env.ATLAS_API_KEY });

function assert(cond: unknown, msg: string): void {
  if (!cond) {
    console.error("FAIL:", msg);
    process.exit(1);
  }
}

const PRINCIPAL = "spiffe://domain-a.test/workload/payments-api";
const DELEGATE = "spiffe://domain-b.test/agent/booking-worker";

const healthy = await atlas.health();
if (!healthy) {
  console.error(`atlas-server not reachable at ${url} — start it with:  go run ./cmd/atlas-server`);
  process.exit(2);
}
console.log("health          ok");
console.log("ready           " + (await atlas.ready() ? "ok" : "warming"));

const grant = await atlas.issue({ principal: PRINCIPAL, delegate: DELEGATE, scope: ["read:orders", "write:audit"] });
assert(grant.record && grant.instance, "issue returned a record + instance");
console.log(`issue           ${grant.instance}`);

const v1 = await atlas.verify(grant.record);
assert(v1.decision === Decision.Accept, `fresh record verifies Accept (got ${v1.decision})`);
assert(v1.trace.length === 5, `trace has five checks (got ${v1.trace.length})`);
console.log(`verify          ${v1.decision}  (${v1.latencyMicros}µs, ${v1.trace.length} checks)`);

await atlas.revoke(grant.instance);
console.log("revoke          ok");

const v2 = await atlas.verify(grant.record);
assert(v2.decision === Decision.Reject, `revoked record verifies Reject (got ${v2.decision})`);
assert(v2.causes.some((c) => /Revoked/i.test(c)), `reject cause mentions revocation (got ${JSON.stringify(v2.causes)})`);
console.log(`verify          ${v2.decision}  (${v2.causes.join(", ")})`);

console.log("\nALL PASS — TypeScript SDK works end-to-end against the real engine.");
